text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+
+# Drawbacks
+
+Any change to a pattern as fundamental as file naming will incur some mental
+friction for developers who are accustomed to the current conventions. It is
+hoped that tooling like Robert's migrator and Ember Watson can lessen this
+friction by automating transitions, and that updated guides, generators, and
+blueprints can make these conventions easy to follow.
+
+Of course, we won't prevent usage of the currently used patterns for some time,
+but they will eventually be deprecated. Some efficiencies, especially in the
+resolver, may not be fully realized until the new patterns are used throughout
+a project.
+
+# Alternatives
+
+## The Module Normalization RFC
+
+Perhaps the most prominent alternative that has been explored is the
+[Module Normalization RFC](https://github.com/emberjs/rfcs/pull/124). Module
+Unification shares many aspects with Module Normalization, but with one
+fundamental difference: buckets in Module Normalization are normalized away
+for the resolver, while collections in Module Unification play an important
+role in module resolution.
+
+The Ember Core Team decided that the sleight of hand required to allow buckets
+to be used for organization only, and not for resolution, could create
+confusion. Essentially, modules could conflict across buckets, because they
+could have matching namespaces, names, and types. This kind of conflict could
+not be allowed, so developers would need to understand too much about the
+resolution strategy to make it ergonomic.
+
+## Other Alternatives
+
+[A large number of other alternatives have been explored](https://gist.github.com/dgeb/396fed953184acb04f4f)
+before settling on this recommendation. Feel free to explore the history of any
+of the linked gists to understand some of the subtle alternatives.
+
+Of course, one alternative is to simply not change anything and accept the
+drawbacks discussed in the Motivation section above. However, even if we accept
+inefficiencies in our resolver and confusion over divergent file structuring
+strategies, we still need to solve the "local lookup" problem, which does not
+have a clean solution in today's module system.
+
+# Unresolved questions
+
+## How should tests be co-located in `src`?
+
+Should tests be allowed within `src` via `*-test` types (e.g.
+`component-integration-test`, `component-unit-test`, etc.) within respective
+collections?
+
+If this RFC is approved, then Robert Jackson plans to adapt the
+[Grand Testing Unification RFC](https://github.com/emberjs/rfcs/pull/119) to
+propose answers to these questions.
+
+## What about routable components?
+
+Should routable components have a type that's unique from other components?
+
+Should they exist alongside `route` and `template` types in the `routes`
+collection?
+
+It seems plausible that routable components could simply use the `component`
+type, and that we could lint against allowing template-invocable components
+alongside routes.
+
```

Comment:
Using a filename of `routable-component.js` would be unfortunate IMO.  In the current proposal `component.js` within the route namespace (i.e. `src/ui/routes/posts/component.js`) is not allowed.  This leaves it up to us when we circle back around to the routable components feature RFC to decide what is the best course of action.  

I think the critical point here is that we are not being future incompatible (i.e. making routable components not possible) with this RFC.  I believe that this RFC has done its due diligence here...